core/src/main/java/org/apache/struts2/interceptor/ActionMappingParametersInterceptor.java

```diff
@@ -98,9 +100,14 @@ protected HttpParameters retrieveParameters(ActionContext ac) {
      *                  </p>
      */
     @Override
-    protected void addParametersToContext(ActionContext ac, Map<String, ?> newParams) {
+    protected void addParametersToContext(ActionContext ac, Map<String, Parameter> newParams) {
         HttpParameters previousParams = ac.getParameters();
-        HttpParameters.Builder combinedParams = HttpParameters.create().withParent(previousParams).withExtraParams(newParams);
+
+        // have to unwrap from Parameter to Object since HttpParameters.Builder will re-wrap
+        Map<String, Object> unwrappedParams = newParams.entrySet().stream().collect(Collectors.toMap(
+            Map.Entry::getKey, e -> e.getValue().getObject()));
+
+        HttpParameters.Builder combinedParams = HttpParameters.create().withParent(previousParams).withExtraParams(unwrappedParams);
```

Comment:
I would rather go with extending the builder and adding a new method `withParameters` that accepts `Map<String, Parameter>` - this will allow to reuse it in the future.